ta/trusted_keys/entry.c

```diff
@@ -0,0 +1,325 @@
+	attr.content.ref.length = sizeof(huk_key);
+
+	res = TEE_PopulateTransientObject(hkey, &attr, 1);
+	if (res)
+		goto out_key;
+
+	res = TEE_SetOperationKey(crypto_op, hkey);
+	if (res)
+		goto out_key;
+
+	if (mode == TEE_MODE_ENCRYPT) {
+		res = huk_ae_encrypt(crypto_op, in, in_sz, out, out_sz);
+		if (res)
+			EMSG("huk_AE_encrypt failed: returned %#"PRIx32, res);
+	} else if (mode == TEE_MODE_DECRYPT) {
+		res = huk_ae_decrypt(crypto_op, in, in_sz, out, out_sz);
+		if (res)
+			EMSG("huk_AE_decrypt failed: returned %#"PRIx32, res);
+	} else {
+		TEE_Panic(0);
+	}
+
+out_key:
+	TEE_FreeTransientObject(hkey);
+out_op:
+	TEE_FreeOperation(crypto_op);
+	memzero_explicit(huk_key, sizeof(huk_key));
+	return res;
+}
+
+static TEE_Result seal_trusted_key(uint32_t types,
+				   TEE_Param params[TEE_NUM_PARAMS])
+{
+	TEE_Result res = TEE_SUCCESS;
+	uint8_t *in = NULL;
+	uint32_t in_sz = 0;
+	uint8_t *out = NULL;
+	uint32_t out_sz = 0;
+
+	DMSG("Invoked TA_CMD_SEAL");
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+				     TEE_PARAM_TYPE_MEMREF_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	in = params[0].memref.buffer;
+	in_sz = params[0].memref.size;
+	out = params[1].memref.buffer;
+	out_sz = params[1].memref.size;
+
+	if (!in || !in_sz || in_sz > MAX_BUF_SIZE)
+		return TEE_ERROR_BAD_PARAMETERS;
+	if ((!out && out_sz > 0) ||
+	    (out && !ALIGNMENT_IS_OK(out, struct tk_blob_hdr)) ||
+	    out_sz > MAX_BUF_SIZE)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if ((in_sz + sizeof(struct tk_blob_hdr)) > out_sz) {
+		params[1].memref.size = in_sz + sizeof(struct tk_blob_hdr);
+		return TEE_ERROR_SHORT_BUFFER;
+	}
+
+	res = huk_crypt(TEE_MODE_ENCRYPT, in, in_sz, out, &out_sz);
+	if (res == TEE_SUCCESS)
+		params[1].memref.size = out_sz;
+
+	return res;
+}
+
+static TEE_Result unseal_trusted_key(uint32_t types,
+				     TEE_Param params[TEE_NUM_PARAMS])
+{
+	TEE_Result res = TEE_SUCCESS;
+	uint8_t *in = NULL;
+	uint32_t in_sz = 0;
+	uint8_t *out = NULL;
+	uint32_t out_sz = 0;
+
+	DMSG("Invoked TA_CMD_UNSEAL");
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+				     TEE_PARAM_TYPE_MEMREF_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	in = params[0].memref.buffer;
+	in_sz = params[0].memref.size;
+	out = params[1].memref.buffer;
+	out_sz = params[1].memref.size;
+
+	if (!in || !ALIGNMENT_IS_OK(in, struct tk_blob_hdr) ||
+	    in_sz <= sizeof(struct tk_blob_hdr) || in_sz > MAX_BUF_SIZE)
+		return TEE_ERROR_BAD_PARAMETERS;
+	if ((!out && out_sz > 0) || out_sz > MAX_BUF_SIZE)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (in_sz > (out_sz + sizeof(struct tk_blob_hdr))) {
+		params[1].memref.size = in_sz - sizeof(struct tk_blob_hdr);
+		return TEE_ERROR_SHORT_BUFFER;
+	}
+
+	res = huk_crypt(TEE_MODE_DECRYPT, in, in_sz, out, &out_sz);
+	if (res == TEE_SUCCESS)
+		params[1].memref.size = out_sz;
```

Comment:
`assert(out_sz == in_sz - sizeof(struct tk_blob_hdr));`?

Reply:
Sounds like a safe check. Will add corresponding check for both seal and unseal operations.